File: usb_android/lib/src/my_usb_configuration.dart
import 'package:usb_platform_interface/usb_platform_interface.dart';

import 'my_api.dart';
import 'my_object.dart';
import 'my_usb_interface.dart';

final myUsbConfigurationApi = MyUsbConfigurationHostApi();

class MyUsbConfiguration extends MyObject implements UsbConfiguration {
  MyUsbConfiguration(super.hashCode);

  @override
  Future<int> getId() {
    return myUsbConfigurationApi.getId(hashCode);
  }

  @override
  Future<UsbInterface> getInterface(int index) async {
    final usbInterfaceHashCode =
        await myUsbConfigurationApi.getInterface(hashCode, index);
    return MyUsbInterface(usbInterfaceHashCode);
  }

  @override
  Future<int> getInterfaceCount() {
    return myUsbConfigurationApi.getInterfaceCount(hashCode);
  }

  @override
  Future<bool> getIsRemoteWakeup() {
    return myUsbConfigurationApi.getIsRemoteWakeup(hashCode);
  }

  @override
  Future<bool> getIsSelfPowered() {
    return myUsbConfigurationApi.getIsSelfPowered(hashCode);
  }

  @override
  Future<int> getMaxPower() {
    return myUsbConfigurationApi.getMaxPower(hashCode);
  }

  @override
  Future<String?> getName() {
    return myUsbConfigurationApi.getName(hashCode);
  }
}

File: usb_android/android/src/main/kotlin/dev/yanshouwang/usb/MyUsbConfiguration.kt
package dev.yanshouwang.usb

import android.hardware.usb.UsbConfiguration

class MyUsbConfiguration(private val collector: MyCollector) : MyUsbConfigurationHostApi {
    override fun getId(hashCode: Long): Long {
        val usbConfiguration = collector.instanceOf(hashCode) as UsbConfiguration
        return usbConfiguration.id.toLong()
    }

    override fun getName(hashCode: Long): String? {
        val usbConfiguration = collector.instanceOf(hashCode) as UsbConfiguration
        return usbConfiguration.name
    }

    override fun getMaxPower(hashCode: Long): Long {
        val usbConfiguration = collector.instanceOf(hashCode) as UsbConfiguration
        return usbConfiguration.maxPower.toLong()
    }

    override fun getInterfaceCount(hashCode: Long): Long {
        val usbConfiguration = collector.instanceOf(hashCode) as UsbConfiguration
        return usbConfiguration.interfaceCount.toLong()
    }

    override fun getIsRemoteWakeup(hashCode: Long): Boolean {
        val usbConfiguration = collector.instanceOf(hashCode) as UsbConfiguration
        return usbConfiguration.isRemoteWakeup
    }

    override fun getIsSelfPowered(hashCode: Long): Boolean {
        val usbConfiguration = collector.instanceOf(hashCode) as UsbConfiguration
        return usbConfiguration.isSelfPowered
    }

    override fun getInterface(hashCode: Long, index: Long): Long {
        val usbConfiguration = collector.instanceOf(hashCode) as UsbConfiguration
        val usbInterface = usbConfiguration.getInterface(index.toInt())
        collector.allocate(usbConfiguration)
        return usbInterface.hashCode().toLong()
    }
}